### PR TITLE
fix: `str.replace(n=0)` on pandas-like backends; cover replace regex and `n` edge cases in tests

### PR DESCRIPTION
**Pull request by jonasdedden**
# Description

Parametrized `replace_edge_data` plus expr/series tests for the `str.replace` cases that agree across eager backends: literal replacement on non-ASCII text (`ααα-x`, `héllo wörld` - byte-vs-char offsets), a regex whose match text also
occurs earlier (`abcx abc` with `abc\b`), replacement containing the pattern with `n=2` (`aa` -> `abab`), `n=0` unchanged, and empty-pattern prepend. Expr tests xfail for `pyspark`/`duckdb`/`ibis` (replace_all-only), matching existing tests. The Greek-α row carries `# noqa: RUF001` (precedent in `to_uppercase_to_lowercase_test.py`): the confusable characters are the point.

Now covered, pinned to polars with `xfail` for the rest: `$1` capture-group expansion (pandas-likes and pyarrow splice it literally), backslash-one without a group (stays literal on polars, rejected elsewhere), and null replacement values (polars leaves the input unchanged, pandas-likes and pyarrow only accept strings). Regex-engine differences are intentionally not fixed here. The stale "n>1 with regex raises" claim from the gap notes is dropped: it only holds for multivalue replacements.

## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [X] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [x] ✅ Test
- [ ] 🐳 Other

## Related issues

None, test-only follow-up from narwhals-daft differential testing.

## AI assistance

- [ ] No AI tools were used for this PR.
- [x] AI tools were used.

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes (N/A — test-only, no behavior change)
- [ ] If this is your first PR to narwhals, attach a screenshot of `pytest` passing locally (not CI):
  
Full-suite command not run; targeted run on the CI constructor set instead:
  `pytest tests/expr_and_series/str/replace_test.py --constructors=pandas,pandas[nullable],pandas[pyarrow],pyarrow,polars[eager],polars[lazy],duckdb,sqlframe` -> 202 passed, 84 xfailed. Also green with `--constructors=ibis` and `--constructors=dask`.

### CODE REVIEW

File: tests/expr_and_series/str/replace_test.py [hunk 1]
```diff
@@ -159,6 +196,187 @@ def test_str_replace_expr_scalar(
...
+    ("data", "pattern", "value", "n", "literal", "expected"), replace_edge_data
+)
+def test_str_replace_edge_series_scalar(
+    constructor_eager: ConstructorEager,
+    request: pytest.FixtureRequest,
+    data: dict[str, list[str]],
+    pattern: str,
+    value: str,
+    n: int,
+    literal: bool,  # noqa: FBT001
```
Comment (FBruzzesi) [inline tests/expr_and_series/str/replace_test.py]:
```suggestion
    n: int,
    *,
    literal: bool,
```

File: tests/expr_and_series/str/replace_test.py [hunk 2]
```diff
@@ -159,6 +196,187 @@ def test_str_replace_expr_scalar(
...
+    ("data", "pattern", "value", "n", "literal", "expected"), replace_edge_data
+)
+def test_str_replace_edge_expr_scalar(
+    constructor: Constructor,
+    request: pytest.FixtureRequest,
+    data: dict[str, list[str]],
+    pattern: str,
+    value: str,
+    n: int,
+    literal: bool,  # noqa: FBT001
```
Comment (FBruzzesi) [inline tests/expr_and_series/str/replace_test.py]:
```suggestion
    n: int,
    *,
    literal: bool,
```

**Review by FBruzzesi — COMMENTED** [review]
Thanks @jonasdedden , for the (`n=0`, pandas-like) case, can't we add the check as you did in other PRs to make it a no-op?

File: tests/expr_and_series/str/replace_test.py (hunk 2 above)
Comment (jonasdedden) [reply tests/expr_and_series/str/replace_test.py]: Thanks!

File: tests/expr_and_series/str/replace_test.py (hunk 1 above)
Comment (jonasdedden) [reply tests/expr_and_series/str/replace_test.py]: Thanks!

**Comment by jonasdedden** [comment]
> Thanks @jonasdedden , for the (n=0, pandas-like) case, can't we add the check as you did in other PRs to make it a no-op?

Yes, and that cleared some test up too

**Review by FBruzzesi — APPROVED** [review]
Thanks @jonasdedden , I added a tiny commit to apply the same fix to dask 🙏🏼
